src/compiler/utilities.ts

```diff
@@ -517,12 +517,46 @@ namespace ts {
 
         if (isJSDocTypeExpressionOrChild(node)) {
             // strip space + asterisk at line start
-            text = text.replace(/(^|\r?\n|\r)\s*\*\s*/g, "$1");
+            text = text.split(/\r\n|\n|\r/).map(l => trimStringStart(l.replace(/^\s*\*/, ""))).join("\n");
         }
 
         return text;
     }
 
+    /**
+     * Removes the leading and trailing white space and line terminator characters from a string.
+     */
+    export const trimString = !!String.prototype.trim ? ((s: string) => s.trim()) : (s: string) => trimStringEnd(s.replace(/^\s+/g, ""));
+
+    /**
+     * Returns a copy with trailing whitespace removed.
+     */
+    export const trimStringEnd = !!String.prototype.trimEnd ? ((s: string) => s.trimEnd()) : trimEndImpl;
+
+
+    /**
+     * Returns a copy with leading whitespace removed.
+     */
+     export const trimStringStart = !!String.prototype.trimStart ? ((s: string) => s.trimStart()) : (s: string) => s.replace(/^\s+/g, "");
+
+    const isWhitespaceRegex = /\s/;
+
+    /**
+     * https://jsbench.me/gjkoxld4au/1
+     * The simple regex for this, /\s+$/g is O(n^2) in v8.
+     * The native .trimEnd method is by far best, but since that's technically ES2019,
+     * we provide a (still much faster than the simple regex) fallback.
+     */
+    function trimEndImpl(s: string) {
+        let end = s.length - 1;
+        while (end >= 0) {
+            // re.test is measurably faster than str.match - probably because there's no result object allocation
+            if (!isWhitespaceRegex.test(s.charAt(end))) break;

```

Comment:
Would looking up the charCode in a map be faster or slower?

Reply:
There are 17 Unicode whitespace characters, in addition to the ASCII range ones. It's probably better to use the regex, just to avoid thinking about those. `.test` is also really fast compared to a normal `str.match` or `.exec` - it just being a boolean result is good for perf.

Reply:
Couldn't we generate an array of (start,end) pairs like we do for `unicodeESNextIdentifierStart` and `unicodeESNextIdentifierPart`, and use the same mechanism to test a character code?

Reply:
After some [benchmarking](https://jsbench.me/yukp35kv3j/1) the simplest thing to do was use our `isWhiteSpaceLike` function that we already export from the scanner - the simple branching is apparently faster than either the regex or a native Set. That's probably a good thing, given how much it's used inside the scanner. 🐱